File: csharp/demo/demo21a02-condition-variable.cs
/*
 * CONDITION VARIABLES
 */
using System;
using System.Threading;



class Demo21A02 : IRunnable
{
    public void run()
    {
        var conditionVar = new object();

        ThreadStart foo = () =>
        {
            Console.WriteLine("foo is waiting...");

            lock (conditionVar)
            {
                Monitor.Wait(conditionVar);
            }

            Console.WriteLine("foo resumed");
        };

        ThreadStart bar = () =>
        {
            for (int i = 0; i < 3; ++i)
            {
                Thread.Sleep(2000);

                lock (conditionVar)
                {
                    Monitor.Pulse(conditionVar);
                }
            }
        };

        for (int i = 0; i < 3; ++i)
            new Thread(foo).Start();

        new Thread(bar).Start();
    }
}

File: python/demo21a02_condition_variable.py
'''
CONDITION VARIABLES
'''

import time
import threading



condition_var = threading.Condition()



def foo():
    print('foo is waiting...')
    with condition_var:
        condition_var.wait()
    print('foo resumed')



def bar():
    for _ in range(3):
        time.sleep(2)
        with condition_var:
            condition_var.notify()



_ = [ threading.Thread(target=foo).start() for _ in range(3) ]
threading.Thread(target=bar).start()
